File: interactivitycommunication/src/main/java/com/codekul/interactivitycommunication/MainActivity.java
package com.codekul.interactivitycommunication;

import android.content.Context;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    public void goToSports(View view) {

        MainActivity main = this;
        Context con = main;

        Class<SportsActivity> cls = SportsActivity.class;

        Intent intent = new Intent(
                con,
                cls
        );

        EditText edNm = findViewById(R.id.edtNm);

        Bundle bundle = new Bundle();
        bundle.putString("yoNm", edNm.getText().toString());

        intent.putExtras(bundle);

        //startActivity(intent);
        startActivityForResult(intent, 1234);

        Car c = new Car();
        Car cB = new Bmw();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1234) {
            // you are coming back from SportsActivity
            if (resultCode == RESULT_OK) {
                if (data != null) {
                    Bundle bndl = data.getExtras();
                    if (bndl != null) {
                        String opNm = bndl.getString("opNm");
                        ((EditText) findViewById(R.id.edtNm)).setText(opNm);
                    }
                }
            }
        }
    }
}

class Car {

}

class Bmw extends Car {

}

File: broadcastreceivers/src/main/java/com/codekul/broadcastreceivers/MainActivity.kt
package com.codekul.broadcastreceivers

import android.content.Intent
import android.content.IntentFilter
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View

class MainActivity : AppCompatActivity() {

    private val br = MyReceiver()
    private val fltr = IntentFilter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        fltr.addAction("com.codekul.my.action")
        fltr.addAction("android.intent.action.ACTION_POWER_CONNECTED")
        fltr.addAction("android.intent.action.ACTION_POWER_DISCONNECTED")
    }

    fun onOkay(vw : View?) {
        sendBroadcast(Intent("com.codekul.my.action"))
    }

    override fun onStart() {
        super.onStart()
        registerReceiver(br, fltr)
    }

    override fun onStop() {
        unregisterReceiver(br)
        super.onStop()
    }
}
